docarray/utils/find.py

```diff
@@ -176,34 +188,94 @@ class MyDocument(BaseDoc):
+            This is because the multiprocessing backend passes the variable via pickle
+            and works in another process.
+            The passed object and the original object do **not** share the same memory.
+    :param num_worker: the number of parallel workers. If not given, then the number of CPUs
+        in the system will be used.
+    :param show_progress: show a progress bar
+    :param pool: use an existing/external pool. If given, `backend` is ignored and you will
+        be responsible for closing the pool.
     :return: a list of named tuples of the form (DocList, AnyTensor),
+
         where the first element contains the closes matches for each query,
         and the second element contains the corresponding scores.
     """
+
     if descending is None:
         descending = metric.endswith('_sim')  # similarity metrics are descending
 
     embedding_type = _da_attr_type(index, embedding_field)
     comp_backend = embedding_type.get_comp_backend()
-
-    # extract embeddings from query and index
+    # extract embeddings from index
     index_embeddings = _extract_embeddings(index, embedding_field, embedding_type)
-    query_embeddings = _extract_embeddings(query, embedding_field, embedding_type)
-
-    # compute distances and return top results
     metric_fn = getattr(comp_backend.Metrics, metric)
-    dists = metric_fn(query_embeddings, index_embeddings, device=device)
-    top_scores, top_indices = comp_backend.Retrieval.top_k(
-        dists, k=limit, device=device, descending=descending
-    )
 
     results = []
-    for indices_per_query, scores_per_query in zip(top_indices, top_scores):
-        docs_per_query: DocList = DocList([])
-        for idx in indices_per_query:  # workaround until #930 is fixed
-            docs_per_query.append(index[idx])
-        docs_per_query = DocList(docs_per_query)
-        results.append(FindResult(scores=scores_per_query, documents=docs_per_query))
+
+    def _get_result(query: Union[DocList, AnyTensor]):
+        q_embed = _extract_embeddings(query, embedding_field, embedding_type)
+        dists = metric_fn(q_embed, index_embeddings, device=device)
+        top_scores, top_indices = comp_backend.Retrieval.top_k(
+            dists, k=limit, device=device, descending=descending
+        )
+        return top_indices, top_scores
+
+    if batch_size is not None:
+        if batch_size <= 0:
+            raise ValueError(
+                f'`batch_size` must be larger than 0, receiving {batch_size}'
+            )
+        else:
+            batch_size = int(batch_size)
+    else:
+        top_indices, top_scores = _get_result(query)
+        res = []
+        for indices_per_query, scores_per_query in zip(top_indices, top_scores):
+            docs_per_query: DocList = DocList([])
+            for idx in indices_per_query:  # workaround until #930 is fixed
+                docs_per_query.append(index[idx])
+            docs_per_query = DocList(docs_per_query)
```

Comment:
not sure that this line is needed

Reply:
Are you refering to the comment "workaround until #930 is fixed" in line 248?